homeassistant/components/xiaomi_miio/fan.py

```diff
@@ -296,25 +307,36 @@
     ATTR_MODE: "mode",
     ATTR_BUZZER: "buzzer",
     ATTR_CHILD_LOCK: "child_lock",
-    ATTR_TARGET_HUMIDITY: "target_humidity",
     ATTR_LED_BRIGHTNESS: "led_brightness",
-    ATTR_USE_TIME: "use_time",
-    ATTR_HARDWARE_VERSION: "hardware_version",
 }
 
 AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER = {
     **AVAILABLE_ATTRIBUTES_AIRHUMIDIFIER_COMMON,
+    ATTR_TARGET_HUMIDITY: "target_humidity",
+    ATTR_HARDWARE_VERSION: "hardware_version",
+    ATTR_USE_TIME: "use_time",
```

Comment:
Is this a relative time? Relative times are not allowed in the state machine. We only allow absolute utc timestamps.

Reply:
This a refactoring of existing attributes, I just moved some unsupported attributes from `common` to more specific attributes sets. `use_time` attribute was added in 25.03.2018 by @syssi (see https://github.com/iromeo/core/commit/e36f27d6fd5b0655375bb90c66cbe108fe967269). So I assume that it is in UTC.

Reply:
Unfortunately this platform isn't following our design guidelines so we can not trust that the existing code is ok.

Reply:
You recently accepted PR https://github.com/home-assistant/core/pull/31729 with the same code. 

I could revert this change to the state before my commit if you want. But I don't think that copy-paste is better than trivial refactoring of the code working for more than 2 years.

Reply:
We won't accept this PR until the integration and platform has been refactored. That should be done in a separate PR first.

> This integration and platform needs to be refactored first to follow our design guidelines regarding fan modes, state attributes and entity access, before any more additions are made.

Reply:
@rytilahti, @syssi As code owners of the integration do you understand how it should be refactored to address this request?

Reply:
@syssi is more knowledgeable on this specific case, but some things I feel should be refactored:

* Moving supported features, available attributes towards the backend library.
  * This involves, e.g., moving the list of available fan speeds to the backend, like was done recently for the vacuum platform.
  * Homeassistant has only 3 different fan speeds (+ off), so rest of those need to be exposed over separate services (if wanted).

* Moving supported models also towards the backend. This will simplify future developments when the backend lib is the only source of truth.

* I think reading somewhere that sensors are preferred over custom state attributes, but cannot find the reference from the docs at the moment. So, these attributes are the ones that should be exposed through the main device: https://developers.home-assistant.io/docs/entity_fan/#properties
  * This would mean creating new entities for those custom attributes, that are linked to the main device.
  * We can adapt the backend to report what type of attribute something is, that can then be used inside homeassistant to create sensors, binary_sensors, switches, etc. depending on the type.
  * I'm not sure what Martin means with entity access, maybe that's related?

Reply:
Ah, found it (thanks @MartinHjelmare on that other PR: https://developers.home-assistant.io/docs/entity_sensor/#available-device-classes . 

So in this case, to my understanding, we shuld have the following:
* sensors:
  * temperature
  * humidity

* binary_sensors:
  * child lock
  * led
  * no water
  * lid open
  * buzzer
  
* unknown:
  * mode
  * led brightness

Reply:
One more point, this architecture issue is relevant to this, too: https://github.com/home-assistant/architecture/issues/310

Reply:
@rytilahti I think that's a great plan and summary. It's not a small task though, so it will take some time and need to be taken step wise. :+1: 

With entity access I meant accessing the entity directly from outside the entity instead of via core or dispatcher. Direct entity access from outside is fragile since it only works after the entity has been added to home assistant. Now we also have a service helper that platforms can use to easily register services that need to call into an entity method.

https://github.com/home-assistant/core/blob/0793b5ac6254538c499ad69906d46fde9eeec2d2/homeassistant/components/xiaomi_miio/fan.py#L573-L574

Here's the docs for the service helper:
https://developers.home-assistant.io/docs/dev_101_services#entity-services

And here's a real example how to use the service helper:
https://github.com/home-assistant/core/blob/34ecc803b312cb2d3b8872db20cd4b8c51cfe4d7/homeassistant/components/wled/light.py#L56-L71